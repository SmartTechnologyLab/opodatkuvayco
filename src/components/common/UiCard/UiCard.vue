<script setup lang="ts">
import Card from 'primevue/card'
</script>

<template>
  <Card class="card">
    <template #title>
      <slot name="title" />
    </template>
    <template #content>
      <slot />
    </template>
  </Card>
</template>

<style scoped lang="scss">
$card-background: $main-text-color;

.card {
  max-width: 760px;
  background: $card-background;
  color: black;
}
</style>
