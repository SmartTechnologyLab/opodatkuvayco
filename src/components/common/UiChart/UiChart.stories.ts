import type { Meta, StoryFn } from '@storybook/vue3'
import UiChart from './UiChart.vue'
import { setChartData } from './common/helpers'

type StoryComponent = StoryFn<typeof UiChart>

const chartData = setChartData([65, 59, 80, 81, 56], ['January', 'February', 'March', 'April', 'May'])

const meta: Meta<typeof UiChart> = {
  component: UiChart,
  args: {
    chartData
  },
  argTypes: {
    chartData: {
      control: 'object'
    }
  }
}

export default meta

export const Bar: StoryComponent = (args) => ({
  components: { UiChart },
  setup() {
    return { args }
  },
  template: `
    <div>
      <UiChart v-bind="{ ...args }" type="bar" />
    </div>
  `
})

export const Doughnut: StoryComponent = (args) => ({
  components: { UiChart },
  setup() {
    return { args }
  },
  template: `
    <UiChart v-bind="{ ...args }" type="doughnut" />
  `
})

export const Line: StoryComponent = (args) => ({
  components: { UiChart },
  setup() {
    return { args }
  },
  template: `
    <UiChart v-bind="{ ...args }" type="line" />
  `
})

export const Pie: StoryComponent = (args) => ({
  components: { UiChart },
  setup() {
    return { args }
  },
  template: `
    <UiChart v-bind="{ ...args }" type="pie" />
  `
})
