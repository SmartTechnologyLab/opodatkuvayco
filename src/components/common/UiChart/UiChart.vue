<script setup lang="ts">
import { computed, toRefs, useAttrs } from 'vue'
import Chart, { type ChartProps } from 'primevue/chart'
import type { ChartData } from './common/types'
import { mergeRight } from 'ramda'
import { DEFAULT_CHART_OPTIONS } from './common/constants'

const attrs = useAttrs()

const props = defineProps<{
  chartData: ChartData
  options?: ChartProps['options']
}>()

const { options } = toRefs(props)

const mergedOptions = computed(() => mergeRight(DEFAULT_CHART_OPTIONS, options.value))
</script>

<template>
  <Chart :data="chartData" :options="mergedOptions" v-bind="{ ...attrs }" />
</template>

<style scoped lang="scss"></style>
