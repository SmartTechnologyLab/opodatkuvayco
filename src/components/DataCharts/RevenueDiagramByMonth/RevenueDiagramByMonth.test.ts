import { shallowMount, type VueWrapper } from '@vue/test-utils'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import RevenueChartByMonth from '@/components/DataCharts/RevenueDiagramByMonth/RevenueDiagramByMonth.vue'
import { stubComponent } from '@/helpers/testHelpers/createComponent'
import UiCard from '@/components/common/UiCard/UiCard.vue'
import type { Deal, Operation } from '@/components/common/DataTable/mocks'
import UiChart from '@/components/common/UiChart/UiChart.vue'

vi.mock('@/components/DataCharts/common/helpers', () => ({
  getRevenueByPurchaseMonth: vi.fn().mockImplementation((operation: Operation) => {
    return () => {
      if (operation === 'sale') {
        return {
          labels: ['січень'],
          datasets: [
            {
              label: 'MSFT',
              data: [23166]
            }
          ]
        }
      } else {
        return {
          labels: ['січень'],
          datasets: [
            {
              label: 'MSFT',
              data: [7321.86]
            }
          ]
        }
      }
    }
  })
}))

const mockDeals: Deal[] = [
  {
    id: Symbol(),
    purchase: {
      commission: 2.7,
      date: new Date('2021-01-10'),
      price: 15,
      rate: 26,
      sum: 270,
      uah: 7321.86
    },
    quantity: 18,
    sale: {
      commission: 5.94,
      date: new Date('2021-01-11'),
      price: 33,
      rate: 39,
      sum: 594,
      uah: 23166
    },
    ticker: 'MSFT',
    total: 15844.14,
    percent: 2.163950143815916
  }
]

const UiCardStub = stubComponent(UiCard, {
  template: `
    <div>
      <slot name="title"></slot>
      <slot />
    </div>
  `
})

describe('RevenueDiagramByMonth', () => {
  let wrapper: VueWrapper<InstanceType<typeof RevenueChartByMonth>>

  const getCardTitle = (operation: 'sale' | 'purchase') =>
    `Діаграма ${operation === 'purchase' ? 'придбань' : 'продажів'} за рік`

  beforeEach(() => {
    wrapper = shallowMount(RevenueChartByMonth, {
      global: {
        stubs: {
          UiChart,
          UiCard: UiCardStub
        }
      },
      props: {
        deals: mockDeals,
        operation: 'sale'
      }
    })
  })

  it('should render the sell card title', () => {
    const CARD_TITLE = getCardTitle(wrapper.props().operation)
    const uiCard = wrapper.findComponent(UiCard)

    expect(uiCard.text()).toBe(CARD_TITLE)
  })

  it('should render the purchase card title', async () => {
    const uiCard = wrapper.findComponent(UiCard)
    await wrapper.setProps({ operation: 'purchase' })

    const CARD_TITLE = getCardTitle(wrapper.props().operation)

    expect(uiCard.text()).toBe(CARD_TITLE)
  })

  it('should calculate data when deals passed', () => {
    const uiChart = wrapper.findComponent(UiChart)

    expect(uiChart.props().chartData).toMatchSnapshot()
  })

  it('should change operation and recalculate the result', async () => {
    await wrapper.setProps({ operation: 'purchase' })

    const uiChart = wrapper.findComponent(UiChart)

    expect(uiChart.props().chartData).toMatchSnapshot()
  })
})
