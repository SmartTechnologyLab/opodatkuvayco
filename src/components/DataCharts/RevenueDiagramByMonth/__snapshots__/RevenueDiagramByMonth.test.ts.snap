// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RevenueDiagramByMonth > should calculate data when deals passed 1`] = `
{
  "datasets": [
    {
      "backgroundColor": [
        "",
        "",
      ],
      "data": [
        [
          "січень",
        ],
        [
          {
            "data": [
              23166,
            ],
            "label": "MSFT",
          },
        ],
      ],
    },
  ],
  "labels": [
    "labels",
    "datasets",
  ],
}
`;

exports[`RevenueDiagramByMonth > should change operation and recalculate the result 1`] = `
{
  "datasets": [
    {
      "backgroundColor": [
        "",
        "",
      ],
      "data": [
        [
          "січень",
        ],
        [
          {
            "data": [
              7321.86,
            ],
            "label": "MSFT",
          },
        ],
      ],
    },
  ],
  "labels": [
    "labels",
    "datasets",
  ],
}
`;
